Fix dockerfile parser failing silently on long tokens

dnephin opened the pull request:
Fixes docker/for-linux#157
Fixes docker/for-mac#2208

Report errors when the scanner fails. The error message should include enough details for the user to fix the problem.

Fixing the first problem revealed another problem with handling remote urls that point at Dockerfiles, which is fixed in the second and third commits.

dnephin committed:
Fix dockerfile parser failing silently on long tokens

Signed-off-by: Daniel Nephin <dnephin@docker.com>

vrothberg commented on builder/dockerfile/parser/parser_test.go:
```diff
@@ -159,3 +160,14 @@ RUN indented \
 }
+
+func TestParseReturnsScannerErrors(t *testing.T) {
+	label := strings.Repeat("a", 65536)
```
For the sake of consistency, I suggest using `bufio.MaxScanTokenSize` instead.

dest2013 replied:
F

vrothberg commented on builder/dockerfile/parser/parser.go:
```diff
@@ -358,3 +358,12 @@ func processLine(d *Directive, token []byte, stripLeftWhitespace bool) ([]byte,
+func handleScannerError(err error) error {
+	switch err {
+	case bufio.ErrTooLong:
+		return errors.Errorf("dockerfile line greater than max allowed size of %d", bufio.MaxScanTokenSize)
```
The error occurs when size is greater *or equal*.

dnephin committed:
Fix remote build target as Dockerfile

The test was passing previously because the preamble was already buffered. After
the change to return Scanner.Err() the final read error on the buffer was no
longer being ignored.

Signed-off-by: Daniel Nephin <dnephin@docker.com>

tonistiigi commented on builder/remotecontext/remote.go:
```diff
@@ -110,7 +86,7 @@ func GetWithStatusError(address string) (resp *http.Response, err error) {
 //    - an error value which will be non-nil either when something goes wrong while
 //      reading bytes from r or when the detected content-type is not acceptable.
-func inspectResponse(ct string, r io.Reader, clen int64) (string, io.ReadCloser, error) {
+func inspectResponse(ct string, r io.Reader, clen int64) (string, io.Reader, error) {
```
Would it be cleaner for this function just to take a `ReadCloser` and return `ReadCloser` so caller doesn't need to do any wrapping?

dnephin replied:
I guess you mean move the `ioutils.NewReadCloserWrapper()` into `inspectResponse()` ? I think either way it needs to be wrapped, it's just about where it happens.

That could work. It would mean that the two error returns in this function would have to call `r.Close()` instead of the one error case in the caller calling closer.

It also seems like it is more the responsibility of `downloadRemote()` since it is aware that it's a response body, where as `inspectResponse` only cares about a reader.

I have a very slight preference for the current implementation, but if you think it's cleaner to move it to inspectResponse I can do that.

I can also remove `closer()` that doesn't need to exist anymore, it can just be `response.Body.Close`

tonistiigi replied:
ok, lets leave it like it is atm

tonistiigi reviewed (commented):
Could you squash 2nd and 3rd commit

dnephin commented:
> Could you squash 2nd and 3rd commit

Yes, definitely. Just wanted to make sure you were happy with the final result before squashing. Will do that now.

tonistiigi reviewed (approved):
LGTM

thaJeztah reviewed (approved):
LGTM

thaJeztah commented:
ping @tiborvass

dnephin commented:
Ok to merge this?

thaJeztah commented:
Ah, yes, I think so 👍

disarticulate commented:
Is this bug elsewhere?

I'm getting a null response on something like

```
"Entrypoint":[
	"/gnatsd",
	"-DV",
	"--tls",
	"--tlskey=/run/secrets/meleorkey",
	"--tlscert=/run/secrets/meleorcrt",
	"--tlscacert=/run/secrets/rootca",
	"--auth=SuperSecretAuthorizationIsForFoolsAndFillers"
],
```

thaJeztah commented:
@disarticulate this fix is included in Docker 17.12 and up
